yellow-paper/docs/addresses-and-keys/addresses-and-keys.md

```diff
@@ -0,0 +1,903 @@
+- _"[The nullifier secret key's (nsk's)] purpose in Sapling was as defense-in-depth, in case RedDSA [(the scheme used for signing txs, using the authentication secret key ask)] was found to have weaknesses; an adversary who could recover ask would not be able to spend funds. In practice it has not been feasible to manage nsk much more securely than a full viewing key [(dk, ak, nk, ovk)], as the computational power required to generate Sapling proofs has made it necessary to perform this step [(deriving nk from nsk)] on the same device that is creating the overall transaction (rather than on a more constrained device like a hardware wallet). We are also more confident in RedDSA now."_
+
+A nullifier public key might have the benefit (in Aztec) that a user could (optionally) provide their nullifier key nk to some 3rd party, to enable that 3rd party to see when the user's notes have been nullified for a particular app, without having the ability to nullify those notes.
+    - This presumes that within a circuit, the nk (not a public key; still secret!) would be derived from an nsk, and the nk would be injected into the nullifier.
+    - BUT, of course, it would be BAD if the nk were derivable as a bip32 normal child, because then everyone would be able to derive the nk from the master key, and be able to view whenever a note is nullified!
+    - The nk would need to ba a hardened key (derivable only from a secret).
+
+Given that it's acceptable to ZCash Orchard, we accept that a nullifier master secret key may be 'seen' by Aztec software.
+    
+
+### Diversified Addresses
+
+This is perhaps a "nice to have", although it's a core feature of zcash keys.
+
+Nice to have:
+- Alice can derive a diversified address; a random-looking address which she can (interactively) provide to Bob, so that Bob may send her funds (and general notes).
+  - Reason: By having the recipient derive a distinct payment address _per counterparty_, and then interactively provide that address to the sender, means that if two counterparties collude, they won't be able to convince the other that they both interacted with the same recipient.
+- Distributing many diversified addresses must not increase the amount of time needed to scan the blockchain (they must all share a single set of viewing keys).
+
+
+### Stealth Addresses
+
+Not to be confised with diversified addresses. A diversified address is generated by the recipient, and interactively given to a sender, for the sender to then use. But a stealth address is generated by the _sender_, and non-interactively shared with the recipient.
+
+Requirement:
+- Random-looking addresses can be derived from a 'main' address, so that private -> public function calls don't reveal the true `msg_sender`. These random-looking addresses can be provably linked back to the 'main' address. 
```

Comment:
I'm having a hard time wrapping my head around this. Who should be able to prove this link?

Reply:
It'll depend on the context. A user can create a stealth/diversified address for themselves (which is provably linked to their main address), and use that address in the public world. Also, someone else can create a 'provably-linked' stealth address for a user, for use in the public (or private) world. Diversified addresses can't be provably-linked to the main account by other people; only by the owner (which is by design) - and so might only be of use in applications which don't want to enforce correct message delivery etc.